src/iterative/KOSMAPOSL/KOSMAPOSLReconstruction.cxx

```diff
@@ -0,0 +1,1487 @@
+//
+//
+/*
+
+    Copyright (C) 2000 PARAPET partners
+    Copyright (C) 2000 - 2011-12-31, Hammersmith Imanet Ltd
+    Copyright (C) 2012-06-05 - 2012, Kris Thielemans
+    Copyright (C) 2018 Commonwealth Scientific and Industrial Research Organisation
+    Copyright (C) 2018 University of Leeds
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+
+/*!
+  \file
+  \ingroup KOSMAPOSL
+  \ingroup reconstructors
+  \brief  implementation of the stir::KOSMAPOSLReconstruction class
+
+  \author Daniel Deidda
+  \author Ashley Gillman
+  \author Matthew Jacobson
+  \author Sanida Mustafovic
+  \author Kris Thielemans
+  \author PARAPET project
+      
+*/
+
+#include "stir/KOSMAPOSL/KOSMAPOSLReconstruction.h"
+#include "stir/OSMAPOSL/OSMAPOSLReconstruction.h"
+#include "stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h"
+#include "stir/DiscretisedDensity.h"
+//#include "stir/LogLikBased/common.h"
+#include "stir/ThresholdMinToSmallPositiveValueDataProcessor.h"
+#include "stir/ChainedDataProcessor.h"
+#include "stir/Succeeded.h"
+#include "stir/numerics/divide.h"
+#include "stir/thresholding.h"
+#include "stir/is_null_ptr.h"
+#include "stir/NumericInfo.h"
+#include "stir/utilities.h"
+// for get_symmetries_ptr()
+#include "stir/DataSymmetriesForViewSegmentNumbers.h"
+#include "stir/ViewSegmentNumbers.h"
+#include "stir/stream.h"
+#include "stir/info.h"
+#include "stir/VoxelsOnCartesianGrid.h"
+
+//#include "stir/modelling/ParametricDiscretisedDensity.h"
+//#include "stir/modelling/KineticParameters.h"
+
+#include <memory>
+#include <iostream>
+#ifdef BOOST_NO_STRINGSTREAM
+#include <strstream.h>
+#else
+#include <sstream>
+#endif
+
+#include "stir/unique_ptr.h"
+#include <algorithm>
+using std::min;
+using std::max;
+#ifndef STIR_NO_NAMESPACES
+using std::cerr;
+using std::endl;
+#endif
+#include "stir/IndexRange3D.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/IO/write_to_file.h"
+
+START_NAMESPACE_STIR
+
+
+// worker functions
+
+
+namespace { // priave namespace for internal functions
+
+unsigned int ravel_index(int x, int y, int z,
+                         int min_x, int min_y, int min_z,
+                         int max_x, int max_y, int max_z) {
+  unsigned int ravelled_index
```

Comment:
Perhaps we shouldn't allocate a variable here and just return the value? Also, declare the function inline?